File: packages/java/anip-server/src/main/java/dev/anip/server/GrantReservationResult.java
package dev.anip.server;

import dev.anip.core.ApprovalGrant;

/**
 * Discriminated result for {@link Storage#tryReserveGrant}.
 *
 * <p>{@code ok=true} -> {@link #grant} is set with the post-reservation
 * use_count, {@link #reason} is null. {@code ok=false} -> {@link #reason}
 * is one of: {@code grant_not_found}, {@code grant_expired},
 * {@code grant_consumed}; {@link #grant} is null.</p>
 *
 * <p>v0.23. See SPEC.md §4.8 Phase B (atomic CAS reservation).</p>
 */
public record GrantReservationResult(boolean ok, ApprovalGrant grant, String reason) {

    public static GrantReservationResult success(ApprovalGrant grant) {
        return new GrantReservationResult(true, grant, null);
    }

    public static GrantReservationResult failure(String reason) {
        return new GrantReservationResult(false, null, reason);
    }
}

File: packages/csharp/src/Anip.Server/GrantReservationResult.cs
using Anip.Core;

namespace Anip.Server;

/// <summary>
/// Result of <see cref="IStorage.TryReserveGrantAsync"/>. v0.23.
/// SPEC.md §4.8 Phase B — atomic compare-and-set increment of use_count.
/// On failure, <see cref="Reason"/> is one of:
/// <list type="bullet">
///   <item>grant_not_found</item>
///   <item>grant_expired</item>
///   <item>grant_consumed</item>
/// </list>
/// </summary>
public record GrantReservationResult(bool Ok, ApprovalGrant? Grant, string? Reason)
{
    public static GrantReservationResult Success(ApprovalGrant grant) =>
        new(true, grant, null);

    public static GrantReservationResult Failure(string reason) =>
        new(false, null, reason);
}
